drivers/sensor/lm75/lm75.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2021 Leica Geosystems AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT lm75
+
+#include <device.h>
+#include <devicetree.h>
+#include <drivers/i2c.h>
+#include <drivers/sensor.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(LM75, CONFIG_SENSOR_LOG_LEVEL);
+
+
+#define LM75_REG_TEMP   0x00
+#define LM75_REG_CONFIG 0x01
+#define LM75_REG_T_HYST 0x02
+#define LM75_REG_T_OS   0x03
+
+struct lm75_data {
+	int16_t temp; /*temp in 0.1°C*/
+};
+
+struct lm75_config {
+	const struct device *i2c_dev;
+	uint8_t i2c_addr;
+};
+
+static inline int lm75_reg_read(struct lm75_config *cfg, uint8_t reg,
+				uint8_t *buf, uint32_t size)
+{
+	return i2c_burst_read(cfg->i2c_dev, cfg->i2c_addr, reg, buf, size);
+}
+
+static inline int lm75_fetch_temp(struct lm75_config *cfg, struct lm75_data *data)
+{
+	int ret;
+	uint8_t temp_read[2];
+	int16_t temp;
+
+	ret = lm75_reg_read(cfg, LM75_REG_TEMP, temp_read, sizeof(temp_read));
+	if (ret) {
+		LOG_ERR("Could not fetch temperature [%d]", ret);
+		return -EIO;
+	}
+
+	/* temp is in two's complement.
+	 *  bit 7 in the lower part corresponds to 0.5
+	 */
+	temp = temp_read[0] << 8 | temp_read[1];
+
+	/* shift right by 7, multiply by 10 to get 0.1° and divide by 2 to get °C */
+	data->temp = (temp / 128) * 10 / 2;
```

Comment:
Why do you multiply by 10 here and then divide by 10 later in `lm75_channel_get`? Rather than converting to intermediate units of .1 deg C, it would be simpler and less error-prone to store data in raw sensor units here, then do just one conversion in `lm75_channel_get`.

Reply:
Because of the tricky data arrangement in the sensor.
The first byte is the temperature in °C. The second is 0.5°C in the MSB.
I'm multiplying by 10 to get the temperature in units of 0.1°C and divide by two to get in the 0.5°C units from the low byte.

Reply:
Maybe I should elaborate a little bit more...
The `temp / 128` is a shift that preserves the sign. The MSB is in degree Celsius. Shifting it 7 to the right (/128) results in °C multiplied by two. The LSB only contains a single bit value that is either 0.5 ° or zero. We cannot use this information alone, because if the temperature is zero, we don't know if it is -0.5 or +0.5 °. this is why we merge together the MSB and LSB, to preserve the sign. We cannot have 0.5° in an int and that is why I chose to use 0.1°C as a unit.

Reply:
Non-power-of-two divide operations are expensive, especially on SoCs that don't have a hardware divider, so we should try to avoid them when possible. You can do this instead:
```c
int16_t temp = temp[0] | temp[1]; /* signed temperature, LSB=2^-8 deg C */
val->val1 = temp >> 8; /* integer part, sign-extended to 32-bits */
val->val2 = temp[1] * 1000000 >> 8; /* fractional part scaled to uC, always positive */
```

Some day I'd really like to rework fixed point data representation in the sensor subsystem, because scaling the fractional part to decimal is confusing, computationally inefficient, and wastes bits of precision.

Reply:
The fractional part has to be signed too! Otherwise it is impossible to distinguish between 0.5 and -0.5 °C.
0 cannot carry the signing information.
Non-power-of-two divisions are rarely expensive, because the compiler can substitute them with multiplication in many cases (if the divisor is a constant). https://gmplib.org/~tege/divcnst-pldi94.pdf
Need to check in compiler explorer still.

Reply:
Here we go, no division, only small efficient code:
https://compiler-explorer.com/z/q1fW9oe79

Reply:
-0.5 can be represented as -1 (integer part) + 0.5 (fractional part). The API is not clear about whether this is allowed, although I think it should be.

I don't think it's a great idea to use extra arithmetic operations and then assume compiler optimizations will reduce them, but I also don't think it's worth blocking this PR over.